### PR TITLE
Implement QueueLength in HWDiskStore and populate data in Linux, Free…

### PR DESCRIPTION
**Pull request by Space2Man**
…BSD, Solaris and Windows.

### CODE REVIEW

File: oshi-core/src/main/java/oshi/hardware/HWDiskStore.java [hunk 1]
```diff
@@ -197,6 +201,13 @@ public long getWriteBytes() {
         return this.writeBytes;
     }
 
+    /**
+     * @return the lenght of the disk queue (#I/O's in progress)
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/HWDiskStore.java]: typo :)

Comment (dbwiddis) [reply oshi-core/src/main/java/oshi/hardware/HWDiskStore.java]: Is this instantaneous or average?

File: oshi-core/src/main/java/oshi/hardware/platform/unix/solaris/SolarisDisks.java [hunk 2]
```diff
@@ -52,6 +52,7 @@ public static boolean updateDiskStats(HWDiskStore diskStore) {
             diskStore.setWrites(data.writes);
             diskStore.setReadBytes(data.nread);
             diskStore.setWriteBytes(data.nwritten);
+            diskStore.setQueueLength(data.rlentime / 1000000L);
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/platform/unix/solaris/SolarisDisks.java]: Reading the docs this is cumulative run time, which is not the same as queue length.   Should you not use `wcnt` (count of elements in wait state) or `rcnt` (count of elements in run state)?  I'm not sure this stat is consistent with the values you're getting from other OS's

File: oshi-core/src/main/java/oshi/hardware/HWDiskStore.java [hunk 3]
```diff
@@ -119,24 +123,24 @@ public HWDiskStore(String name, String model, String serial, long size, long rea
...
-        case SOLARIS:
-            diskFound = SolarisDisks.updateDiskStats(this);
-            break;
-        case FREEBSD:
-            diskFound = FreeBsdDisks.updateDiskStats(this);
-            break;
-        default:
-            LOG.error("Unsupported platform. No update performed.");
-            break;
+            case WINDOWS:
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/HWDiskStore.java]: any chance you can change your formatter to not indent here?

**Comment by Space2Man** [comment]
Hi, I am not sure about Solaris. For Linux (see [here](https://www.kernel.org/doc/Documentation/iostats.txt) it's 

```
Field  9 -- # of I/Os currently in progress
    The only field that should go to zero. Incremented as requests are
    given to appropriate struct request_queue and decremented as they finish.
```

For FreeBSD you are parsing iostat anyway so it's qlen (see [here](https://www.freebsd.org/cgi/man.cgi?iostat))

`qlen	   transactions	queue length`

And for Windows it's described [here](https://blogs.technet.microsoft.com/askcore/2012/03/16/windows-performance-monitor-disk-counters-explained/):

```
Current Disk Queue Length
Current Disk Queue Length is a direct measurement of the disk queue present at the time of the sampling. 
```

But for Solaris I am not sure, if it's wcnt or rcnt or wcnt+rcnt. According to [here](https://docs.oracle.com/cd/E23824_01/html/821-1451/spmonitor-4.html) it looks like wcnt is translated as queue length. My feeling says it might be wcnt+rcnt.

What's your view on this?

**Comment by dbwiddis** [comment]
I expect rcnt to be either 0 or 1, while wcnt is only a positive number if rcnt is 1.  It's like going to a supermarket and asking how long the checkout line is.  The rcnt value is whether someone is actively checking out, and wcnt is those people waiting.   Essentially, "everyone being processed" is rcnt+wcnt and "those waiting" is just wcnt.

All that to say... match it up with what the other stats report!

**Comment by Space2Man** [comment]
This is how OpenIndiana is doing it ([kr_iostat.c](https://github.com/OpenIndiana/oi-userland/blob/1ed47a2077bfb7e671c15a56afc3d9451fba23d7/components/sysutils/net-snmp/sun/agent/modules/healthMonitor/kr_iostat.c)):

```
	avw	= (double)DISK_DELTA(wlentime) / hr_etime;
		/* average number of transactions waiting */

	avr	= (double)DISK_DELTA(rlentime) / hr_etime;
		/* average number of transactions running */
```

**Comment by Space2Man** [comment]
The problem is, for Linux and FreeBSD  those are average values (I would guess) but Solaris we only have wcnt (which is only the current value) or the wlentime (which is accumulative since boot if I read that correctly). So we  would need a delta calculation / time diff ...

**Comment by dbwiddis** [comment]
Welcome to the world of cross-platform coding.   :)       What does the user really need?

**Comment by Space2Man** [comment]
Or we need to parse iostat on Solaris, which does the average handling ...

**Comment by dbwiddis** [comment]
The bigger picture is to ask what's valuable from a monitoring perspective.  Cumulative stats are generally good (and you can get the difference to calculate rate).  Instantaneous stats (queue length) are less useful.

**Comment by Space2Man** [comment]
pfff ... just read that stuff again ... on Windows "Current Disk Queue Length" also seems to just be current value. And on Linux as well:

```
Field  9 -- # of I/Os currently in progress
    The only field that should go to zero. Incremented as requests are
    given to appropriate struct request_queue and decremented as they finish.
```

For the cumulative value I think Field 11 makes more sense:

```
Field 11 -- weighted # of milliseconds spent doing I/Os
    This field is incremented at each I/O start, I/O completion, I/O
    merge, or read of these stats by the number of I/Os in progress
    (field 9) times the number of milliseconds spent doing I/O since the
    last update of this field.  This can provide an easy measure of both
    I/O completion time and the backlog that may be accumulating.
```

Need to check if this kind of stat is also available on the other OS.

**Comment by Space2Man** [comment]
pfff ... difficult if you don't have any test systems. But I think the instantaneous values deliver no real value. Either it has to be cumulative or averaged (for the last x seconds) ...

**Pull request closed without merging**